crates/m3/src/gadgets/hash/keccak/lookedup.rs

```diff
@@ -0,0 +1,759 @@
+		let table_witness = witness.init_table(table_id, N_ROWS).unwrap();
+		let mut segment = table_witness.full_segment();
+
+		let trace = trace::keccakf_trace(StateMatrix::default());
+		rb.populate(&mut segment, &[trace]).unwrap();
+
+		let counts = tally(&cs, &mut witness, &[], lookup_chan, &BitAndIndexedLookup).unwrap();
+		// Fill the lookup table with the sorted counts
+		let sorted_counts = counts
+			.into_iter()
+			.enumerate()
+			.sorted_by_key(|(_, count)| Reverse(*count))
+			.collect::<Vec<_>>();
+
+		witness
+			.fill_table_parallel(&bitand_lookup, &sorted_counts)
+			.unwrap();
+		let statement = Statement {
+			boundaries: vec![],
+			table_sizes: witness.table_sizes(),
+		};
+		let ccs = cs.compile(&statement).unwrap();
+		let witness = witness.into_multilinear_extension_index();
+
+		binius_core::constraint_system::validate::validate_witness(
+			&ccs,
+			&[],
+			&statement.table_sizes,
+			&witness,
+		)
+		.unwrap();
+	}
+
+	#[test]
+	fn test_permutation() {
+		const N_ROWS: usize = TEST_VECTOR.len();
+
+		let mut cs = ConstraintSystem::new();
+		let lookup_chan = cs.add_channel("lookup_channel");
+		let permutation = cs.add_channel("permutation_channel");
+		let mut lookup = cs.add_table("bitand_lookup");
+		let bitand_lookup = BitAndLookup::new(&mut lookup, lookup_chan, permutation, 20);
+		let mut table = cs.add_table("test");
+		let keccakf = KeccakfLookedup::new(&mut table, lookup_chan);
+
+		let allocator = Bump::new();
+		let table_id = table.id();
+
+		let mut witness = WitnessIndex::<PackedType<OptimalUnderlier, B128>>::new(&cs, &allocator);
+		let table_witness = witness.init_table(table_id, N_ROWS).unwrap();
+		let mut segment = table_witness.full_segment();
+
+		let state_ins = TEST_VECTOR
+			.iter()
+			.map(|&[state_in, _]| StateMatrix::from_values(state_in))
+			.collect::<Vec<_>>();
+
+		keccakf.populate_state_in(&mut segment, &state_ins).unwrap();
+		keccakf.populate(&mut segment).unwrap();
+
+		let state_outs = keccakf
+			.read_state_outs(&segment)
+			.unwrap()
+			.collect::<Vec<_>>();
+		for (i, actual_out) in state_outs.iter().enumerate() {
+			let expected_out = StateMatrix::from_values(TEST_VECTOR[i][1]);
+			if *actual_out != expected_out {
+				panic!("Mismatch at index {i}: expected {expected_out:#?}, got {actual_out:#?}",);
+			}
+		}
+
+		let counts = tally(&cs, &mut witness, &[], lookup_chan, &BitAndIndexedLookup).unwrap();
+		// Fill the lookup table with the sorted counts
+		let sorted_counts = counts
+			.into_iter()
+			.enumerate()
+			.sorted_by_key(|(_, count)| Reverse(*count))
+			.collect::<Vec<_>>();
+
+		witness
+			.fill_table_parallel(&bitand_lookup, &sorted_counts)
+			.unwrap();
+
+		let statement = Statement {
+			boundaries: vec![],
+			table_sizes: witness.table_sizes(),
+		};
+
+		let ccs = cs.compile(&statement).unwrap();
+		let witness = witness.into_multilinear_extension_index();
+
+		binius_core::constraint_system::validate::validate_witness(
+			&ccs,
+			&[],
```

Comment:
nit: would be better to use `statement.boundaries`. Noticed that it's easy to trip up otherwise.